cmd/topicctl/subcmd/rebalance.go

```diff
@@ -0,0 +1,323 @@
+				topicErrorDict[topicConfig.Meta.Name] = err
+				rebalanceTopicProgressConfig.RebalanceError = true
+				log.Errorf("Ignoring topic %v for rebalance. Got error: %+v", topicConfig.Meta.Name, err)
+			}
+
+			// show topic final progress
+			if rebalanceCtxMap.Enabled {
+				progressStr, err := util.DictToStr(rebalanceTopicProgressConfig)
+				if err != nil {
+					log.Errorf("Got error: %+v", err)
+				} else {
+					log.Infof("Progress: %s", progressStr)
+				}
+			}
+		}
+	}
+
+	// audit at the end of all topic rebalances
+	successTopics := 0
+	errorTopics := 0
+	for thisTopicName, thisTopicError := range topicErrorDict {
+		if thisTopicError != nil {
+			errorTopics += 1
+			log.Errorf("topic: %s failed with error: %v", thisTopicName, thisTopicError)
+		} else {
+			successTopics += 1
+		}
+	}
+	log.Infof("Overall rebalance - success topics: %d, error topics: %d", successTopics, errorTopics)
+	
+	// show overall rebalance progress
+	if rebalanceCtxMap.Enabled {
+		progressStr, err := util.DictToStr(util.RebalanceProgressConfig{
+			SuccessTopics: successTopics,
+			ErrorTopics: errorTopics,
+			ClusterName:        clusterConfig.Meta.Name,
+			ClusterEnvironment: clusterConfig.Meta.Environment,
+			ToRemove:           rebalanceConfig.brokersToRemove,
+		})
+		if err != nil {
+			log.Errorf("Got error: %+v", err)
+		} else {
+			log.Infof("Progress: %s", progressStr)
+		}
+	}
+
+	return nil
+}
+
+// Check whether a topic is a candidate for action rebalance
+// - consistency of topic with cluster config
+// - settings(partitions, retention time) of topic config with settings for topic in the cluster
+func rebalanceTopicCheck(
+	topicConfig config.TopicConfig,
+	clusterConfig config.ClusterConfig,
+	topicInfo admin.TopicInfo,
+) error {
+	// topic config should be same as the cluster config
+	if err := config.CheckConsistency(topicConfig, clusterConfig); err != nil {
+		return err
+	}
+
+	log.Infof("Check topic partitions...")
+	if len(topicInfo.Partitions) != topicConfig.Spec.Partitions {
+		return fmt.Errorf("Topic partitions in kafka does not match with topic config file")
+	}
+
+	log.Infof("Check topic retention.ms...")
+	if topicInfo.Config["retention.ms"] != strconv.Itoa(topicConfig.Spec.RetentionMinutes*60000) {
+		return fmt.Errorf("Topic retention in kafka does not match with topic config file")
+	}
+
+	return nil
+}
+
+// Perform rebalance on a topic. returns error if unsuccessful
+// topic will not be rebalanced if
+//   - topic config is inconsistent with cluster config (name, region, environment etc...)
+//   - partitions of a topic in kafka cluster does not match with topic partition setting in topic config
+//   - retention.ms of a topic in kafka cluster does not match with topic retentionMinutes setting in topic config
+//
+// to ensure there are no disruptions to kafka cluster
+//
+// NOTE: topic that is not present in kafka cluster will not be applied
+func rebalanceApplyTopic(
```

Comment:
https://github.com/segmentio/topicctl/blob/dd82be7c0b175352de33c63d6b40ac5269bb745a/pkg/apply/apply.go#LL719C46-L719C46
Is specific to rebalance assignments

But in reality, updateBalance is what we need. 
https://github.com/segmentio/topicctl/blob/dd82be7c0b175352de33c63d6b40ac5269bb745a/pkg/apply/apply.go#L701

That is achieved with apply

can we resolve this?

Reply:
As discussed earlier, this is wrapper on top of apply

Reply:
If we create a new package and move updateBalance, It will create confusion. updateBalanace is in right location pkg/apply

We do not want to touch existing package layout. I didnt evaluate this yet but it can be disruptive or test cases can failure? It will definitely create confusion to users who might be using existing pkg structure

NOTE: As mentioned earlier, rebalance is just a wrapper on top of apply. 

Until we do not have a separate complete rebalance apply logic (Which it self will need bunch of pkg code to be moved), I am not sure why we need to relocate/move functions to rebalance package

Can this be resolved?